syntax = "proto3";
import "google/protobuf/empty.proto";

option java_multiple_files = true;
option objc_class_prefix = "HLW";

package indevicecommands;

service InDeviceCommands {
  rpc KeepAlive(google.protobuf.Empty) returns (google.protobuf.Empty) {}
  rpc UpdateSender(Connection) returns (google.protobuf.Empty) {}
  rpc UpdateBatchSize(Int32) returns (google.protobuf.Empty) {}
  rpc UpdateResolution(Dimensions) returns (google.protobuf.Empty) {}
  rpc UpdateTimeKeeping(TimeKeeping) returns (google.protobuf.Empty) {}
  rpc SyncDatasources(Int32) returns (google.protobuf.Empty) {}
  rpc SetStartFrame(Int32) returns (google.protobuf.Empty) {}
  rpc FederatedLearningReturn(FlData) returns (google.protobuf.Empty) {}
  rpc RetrieveContainerMetrics(google.protobuf.Empty) returns (ContainerMetrics) {}
  rpc StopExecution(ContainerSignal) returns (google.protobuf.Empty) {}
}

message ContainerSignal {
  string name = 1;
  bool forced = 2;
}

message Connection {
  int32 mode = 1;
  string name = 2;
  string ip = 3;
  int32 port = 4;
  float data_portion = 5;
  string old_link = 6;
  int64 timestamp = 7;
  int32 offloading_duration = 8;
}

message TimeKeeping {
  string name = 1;
  int32 slo = 2;
  uint64 time_budget = 4;
  uint64 start_time = 5;
  uint64 end_time = 6;
  uint64 local_duty_cycle = 7;
  uint64 cycle_start_time = 8;
}

message Dimensions {
  int32 channels = 1;
  int32 height = 2;
  int32 width = 3;
}

message Int32 {
  int32 value = 1;
}

message ContainerMetrics {
  float arrival_rate = 1;
  int32 queue_size = 2;
  int64 avg_latency = 3;
  int32 drops = 4;
  int32 throughput = 5;
}

message FlData {
  string name = 1;
  string device_name = 2;
  uint32 state_size = 3;
  uint32 resolution_size = 4;
  uint32 max_batch = 5;
  uint32 threading_size = 6;
  double latest_loss = 7;
  bytes network = 8;
  optional bytes states = 9;
  optional bytes resolution_actions = 10;
  optional bytes batching_actions = 11;
  optional bytes scaling_actions = 12;
}